simtools/corsika/corsika_output.py

```diff
@@ -0,0 +1,1002 @@
+
+        for step in range(self.num_of_hist):
+            boost_axes_position = self._create_regular_axes("hist_position")
+            self.hist_position.append(
+                bh.Histogram(
+                    boost_axes_position[0],
+                    boost_axes_position[1],
+                    boost_axes_position[2],
+                )
+            )
+            boost_axes_direction = self._create_regular_axes("hist_direction")
+            self.hist_direction.append(
+                bh.Histogram(
+                    boost_axes_direction[0],
+                    boost_axes_direction[1],
+                )
+            )
+
+            boost_axes_time_altitude = self._create_regular_axes("hist_time_altitude")
+            self.hist_time_altitude.append(
+                bh.Histogram(
+                    boost_axes_time_altitude[0],
+                    boost_axes_time_altitude[1],
+                )
+            )
+
+    def _fill_histograms(self, photons, azimuth_angle=None, zenith_angle=None):
+        """Fill the histograms created by self._create_histogram
+
+        Parameters
+        ----------
+        photons: list
+            List of size M of numpy.void of size (N,8), where M is the number of telescopes in the
+            array, N is the number of photons that reached each telescope. The following information
+             of the Cherenkov photons on the ground are saved: x, y, cx, cy, time, zem, photons,
+             wavelength.
+        azimuth_angle: astropy.Quantity
+            Azimuth angle to rotate the observational plane and obtain it perpendicular to the
+            incoming event. It can be passed in radians or degrees.
+            If not given, no rotation is performed.
+        zenith_angle: astropy.Quantity
+            Zenith angle to rotate the observational plane and obtain it perpendicular to the
+            incoming event. It can be passed in radians or degrees.
+            If not given, no rotation is performed.
+
+        Raises
+        ------
+        IndexError:
+            If the index or indices passed though telescope_index are out of range.
+        """
+
+        hist_num = 0
+        for one_tel_info, photons_info in zip(self._tel_positions, photons):
+
+            if azimuth_angle is None or zenith_angle is None:
+                photon_x, photon_y = photons_info["x"], photons_info["y"]
+            else:
+                photon_x, photon_y = rotate(
+                    photons_info["x"],
+                    photons_info["y"],
+                    azimuth_angle,
+                    zenith_angle,
+                )
+
+            if self.telescope_indices is None:
+                hist_num = 0
+                photon_x = -one_tel_info["x"] + photon_x
+                photon_y = -one_tel_info["y"] + photon_y
+
+            if (
+                one_tel_info in self._tel_positions[self.telescope_indices]
```

Comment:
`self._tel_positions[self.telescope_indices]` is itself a numpy.array, so I think it would be the same.

Reply:
No, the `if` here does a search in the array. If you simply loop over `self._tel_positions[self.telescope_indices]`, then you don't need the `if` at all.

Reply:
sorry, it is still not clear to me what you mean. Do you mean to change this loop `        for one_tel_info, photons_info in zip(self.tel_positions, photons):
` ?

If so, then that is not a good idea, because the current implementation is robust in the sense that it handles both cases (if self.telescope_indices is None and if it is not None). Changing this loop will mean that the case in which self.telescope_indices is None is not going to work anymore.

Reply:
For the record, after discussions during the CTA meeting we decided to modify the strategy a bit and make the telescope indices actual indices instead of `None` when wanting all telescope in one plot. That should simplify some of the open questions. I write it here but it's relevant to a few of the open issue. Waiting for the rewrite and then we can look into the new implementation.

Reply:
As we discussed today, I redefined the flag `telescope_indices` to be really the indices of the telescopes even in case we deal with the photons from all telescopes in a single histogram. To define whether we fill the histograms separately for each telescope or all together, I defined now the flag `single_telescopes`. As agreed, we have dropped the idea of filling a big (4000 x 4000 x 80) histogram due to computation issues.